### DISCUSSION TITLE
A Deep Dive Into Terraform Static Code Analysis Tools: Features and Comparisons | devDosvid blog

### DISCUSSION THREAD

**Discussion by giscus** [vasylenko/devdosvid.blog#99] [General]
# A Deep Dive Into Terraform Static Code Analysis Tools: Features and Comparisons | devDosvid blog

Explore key features and comparisons of top Terraform static code analysis tools to enhance security and compliance in your infrastructure management.

https://devdosvid.blog/2024/04/16/a-deep-dive-into-terraform-static-code-analysis-tools-features-and-comparisons/

**Comment by simar7** [comment]
Hi I'm one of the maintainers of Trivy and really appreciate you putting this together! I just wanted to clarify how Trivy counts the number of checks (policies) as it is different from some of the other scanners mentioned here. 

In Trivy we normalize all of the different providers of IaC, e.g. Terraform, CloudFormation etc. into one common state and run all of our checks (policies) on it. This means we only count the check (policies) once, regardless if it comes from different providers.

To give you an example: You can create an AWS S3 bucket with either Terraform or CloudFormation. The bucket not being encrypted would be one single check (instead of 2 checks) because for Trivy at the end of the day it's just a bucket in AWS.

I hope it makes sense. Feel free to reach out if you have any questions!

**Reply by vasylenko** [reply]
Hi @simar7! Thank you for Trivy and thank you for the feedback! I appreciate the context! 
In the regard of details you shared, would you like to adjust the number of out-of-the-box policies I mentioned for Trivy, if that's the case? Or did you mean that the same number of policies apply to all IaC tools rather than just for Terraform?

**Reply by simar7** [reply]
Hi @vasylenko, what I implied is that Trivy doesn't count a single policy multiple times for each IaC provider (terraform, cloudformation etc.) as we think it's not entirely accurate to do so.

While it may make Trivy look like there are less policies than other tools, we do so in order for them being to be unique and accurate. 

Having said that we certainly don't cover everything today and certainly have room for improvement, which your blog helps us better understand so do appreciate the write up!

**Reply by vasylenko** [reply]
@simar7, thanks for explaining! Now I got it.
Wishing you the best of luck with improving your product! I like your approach to be a comprehensive instrument and cover all kinds of IaC, not only Terraform!

**Reply by ehaselwanter** [reply]
I think it’s worth to add Mondoo to that list: https://mondoo.com/docs/cnspec/supplychain/terraform/

**Reply by vasylenko** [reply]
Hi @ehaselwanter! Thank you for your suggestion! I am thinking about the second part of the review to cover more tools such as Snyk, now Mondoo, and some others. 
Could you please help me clarify if the Mondoo IaC scanner is open-sourced? I could not find its repository :(

**Reply by ehaselwanter** [reply]
@vasylenko yes. it is open source: https://github.com/mondoohq/cnspec

**Reply by vasylenko** [reply]
Awesome, thanks @ehaselwanter!

**Comment by DarwinJS** [comment]
Thanks for the article @vasylenko. I wanted to point out that GitLab uses Kics as its default built-in IaC scanner, so in the Integration Capabilities section, the column "CI/CD Systems" should list GitLab. Also you list that Checkov outputs GitLab SAST, but it also does not have GitLab listed under "CI/CD Systems in that same Integration Capabilities. They also provide integration there: https://www.checkov.io/4.Integrations/GitLab%20CI.html and I have created a slightly more advanced integration (implemented as an include) here: https://gitlab.com/guided-explorations/ci-cd-plugin-extensions/checkov-iac-sast

**Reply by DarwinJS** [reply]
@vasylenko - I actually just took the time to convert my implementation to a GitLab CI Component - so now it has global visibility in the GitLab.com CI component catalog here: https://gitlab.com/explore/catalog/guided-explorations/ci-components/checkov-iac-sast

**Reply by vasylenko** [reply]
Hi @DarwinJS! So glad to see you here! 💟 Thank you for reading my blog and for your comments!

I've added the info about KICS+GitLab to the Integration Capabilities section. As for the Checkov+GitLab, this one is kinda not what I meant to check: the idea of the "CI/CD Systems" feature here is whether CI system supports direct integration through plugins or extensions, eliminating the need for manual downloads or CLI setups. Checkov requires you to write the CI job for GitLab like you would with Docker or CLI invocations. If that makes sense. 

Also, I appreciate your work on components! Very inspiring!

**Reply by DarwinJS** [reply]
I think you may be misunderstanding our new components feature. It's nearly identical to shared GitHub actions.in this case a single include.

**Reply by vasylenko** [reply]
@DarwinJS Yep, I might have misunderstood. But I've read the docs again, and it looks like you're right! I added GitLab to the Checkov row on the table under "Integration Capabilities." The one thing that still confuses me is that it's not Checkov who maintains the integration for GitLab, in contrast with the Action (https://github.com/bridgecrewio/checkov-action).

**Reply by DarwinJS** [reply]
Thanks so much for your openness and collaboration- even when I was so rude as not to lead with thanks on my most recent message 😞 - sorry I was out running errands and rushed the message too much.

I am community maintaining it and their existing docs are from my previously created enterprise grade include.

I have an MR into their docs page for this component and if they wish to maintain the component, it'll be all theirs.

Although building a GitLab component that can run successfully in almost any CI code on the planet has some specific CI builder nuances and best practices that sometimes the software owner does not have time to learn.

I do Chocolatey packaging and it is the same there - the community frequently starts packaging something and if the actual vendor wishes to take over they are welcome to.

Also a good packager might be able to build something better than the vendor and folks then prefer a community alternative package.

**Reply by DarwinJS** [reply]
@vasylenko - any chance you'd be willing to update the Checkov line to say "GitLab CI Component" so readers have an idea where to find the GitLab support?

**Reply by vasylenko** [reply]
Thanks for the additional context, @DarwinJS! I appreciate it! I've added clarification about the CI component but in a slightly different way — with the link to the component but under the table, to allow more space for explanation; I hope you don't mind.

**Reply by DarwinJS** [reply]
Thanks for all your changes! So the question then is why does GitHub have the word "Actuons" after it? GitHub Actions and GitLab CI Components are the same concept :)

**Reply by sandipan1988** [reply]
Hi Guys, please give it a try - https://tfgaurd.com/

**Comment by joshd-emergent** [comment]
Great article!

Doesn't Trivy and tfsec support pre-commit hooks? or I am misunderstanding the git hook feature noted here.

https://github.com/antonbabenko/pre-commit-terraform?tab=readme-ov-file#available-hooks

**Comment by sandipan1988** [comment]
Hi Friend,  You can add this product also in the list, have a look - https://tfgaurd.com/
